airflow/providers/dbt/cloud/CHANGELOG.rst

```diff
@@ -28,6 +28,11 @@
 Changelog
 ---------
 
+main
+.....
+
+In Airflow 2.10.0, we fix the way try_number works, so that it no longer returns different values depending on task instance state.  Importantly, after the task is done, it no longer shows current_try + 1. Thus in 3.8.1 we patch this provider to fix try_number references so they no longer adjust for the old, bad behavior.
+
```

Comment:
Why is this here?
How is this PR affect bdt provider?

Reply:
We are discussing this very thing in here with @kacpermuda and @mobuchowski :

* https://github.com/apache/airflow/pull/39513#discussion_r1596005178

But also I need @dstandish to confirm if the SMTP change https://github.com/apache/airflow/pull/39513#discussion_r1596008489 - seem that the DBT change was good actually - as it handles back-compatibility nicely.

This is the case for which I started to implement #39513 (for 2.9.0 compatibility) and follow-ups for 2.8 (39606) and 2.7 (will be next) will be really helpful, because we are going to run the full suite of provider tests for past versions of airlfow once it is complete, so any kind of changes like that when we change both airflow and provider behaviour will have to be handled either as separate PRs or it will have to pass the back-compatibility tests with supported versions of Airflow. 

So those kind of changes will be caught very early in the process - in the PR that modifies tests of provider to accomodate for changed behaviour of Airflow (and it will need to be handled by the author).

Reply:
BTW. Most likely those kind of tests will have to be done as single PRs, because otherwise it will be difficult to coordinate such change in case provider test implicitly depends on some "airflow" internals - like it was in this case where provider behaviour (and tests) are depending on when try_number gets updated.

Reply:
I am still not clear about this entry. I will hold release for the relevant providers till clarification

Reply:
Hi @eladkal 

This newsfragment is in reference to the change to `airflow/providers/dbt/cloud/utils/openlineage.py` in this PR.

The old code did this:

```
        # generate same run id of current task instance
        parent_run_id = OpenLineageAdapter.build_task_instance_run_id(
            dag_id=task_instance.dag_id,
            task_id=operator.task_id,
            execution_date=task_instance.execution_date,
            try_number=task_instance.try_number - 1,
        )
```

Notice that previously it took the "public" try number attr and minused one.  Presumably, this is because when this bit of code is fired, the task has completed and the try_number has been incremented (thus showing the "wrong" try_number).  So the old code was correcting for the bad behavior of try_number in airflow.

In the present PR, we correct this behavior.  The try_number is now only incremented when a new try is scheduled, and it remains unchanged throughout the task try lifecycle.  The code in this provider which previously hacked around the bad behavior is now no longer necessary, and if left in there, it will give the wrong result.  Hence the introduction of the compat shim `_get_try_number` which is added to the same module.